File: leetcode/[0014]longest-common-prefix.ts
// https://leetcode.com/problems/longest-common-prefix/
function longestCommonPrefix(strs: string[]): string {
    return solution3(strs);
}



// compare first to others: 0 ms
function solution1(strs: string[]): string {
    if (strs.length === 0) return "";

    let first = strs[0].match(/./gv) ?? [];
    for (const other of strs.slice(1,)) {
        while (!other.startsWith(first.join(""))) {
            first.pop();
        }
    }

    return first.join("");
}



// use zip generator: 10 ms
function solution2(strs: string[]): string {
    let a = "";
    for (const xs of zip(strs)) {
        if (!xs.every((x) => x === xs[0])) break;

        a += xs[0];
    }

    return a;
}
function* zip(strs: string[]): Generator<string[]> {
    const iters = strs
        .map((str) => str.match(/./gv) ?? [])
        .map((xs) => xs[Symbol.iterator]());

    while (true) {
        let pool = [];
        for (const it of iters) {
            const { value, done } = it.next();
            if (done) return;

            pool.push(value);
        }

        yield pool;
    }
}



// trie: 13 ms
class TrieNode {
    isEnd = false;
    children = new Map<string, TrieNode>();
}
function solution3(strs: string[]): string {
    const trie = new TrieNode();

    for (const str of strs) {
        const cs = str.match(/./gv) ?? [];
        let n = trie;

        for (const c of cs) {
            if (!n.children.has(c)) n.children.set(c, new TrieNode());
            n = n.children.get(c)!;
        }

        n.isEnd = true;
    }

    let a = "";
    let n = trie;
    while (!n.isEnd && n.children.size === 1) {
        const c = [...n.children.keys()][0];
        a += c;
        n = n.children.get(c)!;
    }

    return a;
}

File: leetcode/[0014]longest-common-prefix.rs
// https://leetcode.com/problems/longest-common-prefix/
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        return solution3(strs);
    }
}



// comp first to others: 0 ms
fn solution1(strs: Vec<String>) -> String {
    let strs = strs
        .into_iter()
        .map(|str| str.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();
    
    let mut first = strs[0].clone();

    for other in strs[1..].iter() {
        while !other.starts_with(&first) {
            first.pop();
        }
    }

    return String::from_iter(first);
}



// comp each char: 0 ms
fn solution2(strs: Vec<String>) -> String {
    let strs = strs
        .into_iter()
        .map(|str| str.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();

    let mut a = String::new();

    'outer:
    for i in 0.. {
        let mut tmp = vec![];

        for str in strs.iter() {
            match str.get(i) {
                Some(&x) => {
                    tmp.push(x);
                }
                None => {
                    break 'outer;
                }
            }
        }
        
        let t = tmp[0];
        if tmp.into_iter().any(|x| x != t) {
            break 'outer;
        }

        a.push(t);
    }

    return a;
}



// trie: 0 ms
fn solution3(strs: Vec<String>) -> String {
    use std::collections::HashMap;
    
    struct TrieNode {
        end: bool,
        children: HashMap<char, Self>,
    }

    impl TrieNode {
        fn new() -> Self {
            return Self {
                end: false,
                children: HashMap::new(),
            }
        }
    }

    let mut trie = TrieNode::new();

    let strs = strs
        .into_iter()
        .map(|str| str.chars().collect::<Vec<char>>())
        .collect::<Vec<Vec<char>>>();
    
    for str in strs.iter() {
        let mut n = &mut trie;

        for &x in str.iter() {
            if !n.children.contains_key(&x) {
                n.children.insert(x, TrieNode::new());
            }

            n = n.children.get_mut(&x).unwrap();
        }
        n.end = true;
    }

    let mut a = String::new();
    let mut n = &mut trie;
    while !n.end && n.children.len() == 1 {
        let k = *n.children.keys().collect::<Vec<&char>>()[0];
        a.push(k);
        n = n.children.get_mut(&k).unwrap();
    }

    return a;
}
